include/deal.II/matrix_free/fe_remote_evaluation.h

```diff
@@ -0,0 +1,1085 @@
+      return ptrs[index - start];
+    }
+
+    /**
+     * Get the number of stored values.
+     */
+    unsigned int
+    size() const
+    {
+      Assert(ptrs.size() > 0, ExcInternalError());
+      return ptrs.back();
+    }
+
+    /**
+     * Fill class from outside.
+     */
+    void
+    reinit(const std::vector<unsigned int> &ptrs_in,
+           const unsigned int               start_in = 0)
+    {
+      ptrs  = ptrs_in;
+      start = start_in;
+    }
+
+  private:
+    /**
+     * This parameter can be used if indices do not start with 0.
+     */
+    unsigned int start = 0;
+    /**
+     * Pointers to data at index.
+     */
+    std::vector<unsigned int> ptrs;
+  };
+
+
+
+  /**
+   * Specialization for `cell_face_pairs=true`.
+   */
+  template <>
+  class FERemoteEvaluationDataView<true>
+  {
+  public:
+    /**
+     * Get a pointer to data at (cell_index, face_number).
+     */
+    unsigned int
+    get_shift(const unsigned int cell_index,
+              const unsigned int face_number) const
+    {
+      Assert(cell_index != numbers::invalid_unsigned_int,
+             ExcMessage("Cell index has to be valid!"));
+      Assert(face_number != numbers::invalid_unsigned_int,
+             ExcMessage("Face number has to be valid!"));
+
+      Assert(cell_start <= cell_index, ExcInternalError());
+
+      AssertIndexRange(cell_index - cell_start, cell_ptrs.size());
+      const unsigned int face_index =
+        cell_ptrs[cell_index - cell_start] + face_number;
+      AssertIndexRange(face_index, face_ptrs.size());
+      return face_ptrs[face_index];
+    }
+
+    /**
+     * Get the number of stored values.
+     */
+    unsigned int
+    size() const
+    {
+      Assert(face_ptrs.size() > 0, ExcInternalError());
+      return face_ptrs.back();
+    }
+
+    /**
+     * Fill class from outside.
+     */
+    void
+    reinit(const std::vector<unsigned int> &cell_ptrs_in,
+           const std::vector<unsigned int> &face_ptrs_in,
+           const unsigned int               cell_start_in = 0)
+    {
+      cell_ptrs  = cell_ptrs_in;
+      face_ptrs  = face_ptrs_in;
+      cell_start = cell_start_in;
+    }
+
+  private:
+    /**
+     * This parameter can be used if cell_indices do not start with 0.
+     */
+    unsigned int cell_start = 0;
+
+    /**
+     * Pointers to first face of given cell_index.
+     */
+    std::vector<unsigned int> cell_ptrs;
+
+    /**
+     * Pointers to data at (cell_index, face_number).
+     */
+    std::vector<unsigned int> face_ptrs;
+  };
+
+} // namespace internal
+
+
+
+/**
+ * A class to fill the fields in FERemoteEvaluationData.
+ * FERemoteEvaluation is thought to be used with another @p FEEvaluationType 
+ * (FEEvaluation, FEFaceEvaluation, or FEPointEvaluation). @p is_face specifies 
+ * if @p FEEvaluationType works on faces.
+ */
+template <typename FEEvaluationType_, bool is_face_>
+class FERemoteEvaluationCommunicator : public Subscriptor
```

Comment:
Why is this class not internal?

Reply:
`FERemoteEvaluationCommunicator` is set up by the user and handed to `FEPointEvaluation`. I think we should avoid that classes and functions, intended to end up in user-code, are `internal`. I am open for discussion though :)